url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/API/plugin/CommandHandler.java

```diff
@@ -0,0 +1,214 @@
+	}
+	
+	/**
+	 * Find a command based on the name.
+	 * This will also look for shortcuts
+	 * @param name
+	 *           The name/shortcut of the command
+	 * @return
+	 *        The {@link Command} object
+	 */
+	public Command find(String name) {
+		for (Command c : commands) {
+			if (name.equalsIgnoreCase(c.getName()))
+				return c;
+			else {
+				for (String shortcut : c.getShortcuts()) {
+					if (name.equalsIgnoreCase(shortcut))
+						return c;
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Have <b>player</b> execute a command. This method will
+	 * check for group permissions, if the <b>player</b> can't
+	 * execute the command, the player will be notified. 
+	 * @param player
+	 *              The player executing the command
+	 * @param command
+	 *               The command the player will use
+	 * @param args
+	 *            Any arguments that will be passed to this command
+	 */
+	public void execute(CommandExecutor player, String command, String[] args) {
+		if (find(command) == null) {
+			if (player != null)
+				player.sendMessage("Command not found!");
+			else
+				_server.Log("Command not found!");
+		}
+		Command c = find(command);
+		if (args.length == 1 && args[0].equals(""))
+			args = new String[0];
+		if(c != null)
+		{
+			if (player != null && !player.getGroup().canExecute(c))
+				player.sendMessage("Sorry, you dont have permission to execute this command!");
+			else {
+				_server.Log(player.getName() + " used /" + c.getName() + arrayToString(args));
+				c.execute(player, args);
+			}
+		}
+	}
+	
+	private String arrayToString(String[] args) {
+		String finals = "";
+		for (String s : args) {
+			finals += " " + s;
+		}
+		return finals;
+	}
+	
+	/**
+	 * Add a command to the command list. This will load permissions
+	 * for this command, if no permissions are found for it, it will
+	 * save the default permissions for it.
+	 * @param cmd
+	 *           The command to add.
+	 */
+	public void addCommand(Command cmd) {
+		if (commands.contains(cmd))
+			return;
+		commands.add(cmd);
+		try {
+			loadPermissions();
+			savePermissions();
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+	
+	/**
+	 * Remove a command from the list
+	 * @param cmd
+	 *           The command to remove
+	 */
+	public void removeCommand(Command cmd) {
+		if (!commands.contains(cmd))
+			return;
+		commands.remove(cmd);
+	}
+	
+	/**
+	 * Remove a command from the list.
+	 * @param name
+	 *            The command name.
+	 */
+	public void removeCommand(String name) {
+		Command c = find(name);
+		if (c == null)
+			return;
+		removeCommand(c);
+	}
+	
+	private void loadPermissions() throws IOException {
+		if (!new File("properties").exists())
+			new File("properties").mkdir();
+		if (!new File("properties/commands.config").exists())
+			makeDefault();
+		FileInputStream fstream = new FileInputStream("properties/commands.config");
+		DataInputStream in = new DataInputStream(fstream);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 156, 157, 159. The resource is closed at line 170. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).